File: services/buyer-client/src/main/java/com/agentpay/buyer/keys/CanonicalForm.java
package com.agentpay.buyer.keys;

import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;

/**
 * Canonical byte form a buyer agent signs over before submitting POST /payments (NFR-S-005). The
 * fields and order are: {@code merchant_id | amount | currency | case_id}. Pipe separators avoid
 * the ambiguity of concatenation. {@code amount} is normalised to {@code toPlainString} (no
 * scientific notation; trailing zeros preserved exactly as they appear on the wire) so a verifier
 * can reproduce the bytes from the JSON request body. {@code case_id} serves as the per-request
 * nonce — there is no separate nonce field on PaymentRequest.
 */
public final class CanonicalForm {

  private CanonicalForm() {}

  public static byte[] bytes(String merchantId, BigDecimal amount, String currency, String caseId) {
    if (merchantId == null || amount == null || currency == null || caseId == null) {
      throw new IllegalArgumentException(
          "canonical-form inputs must all be non-null (merchantId, amount, currency, caseId)");
    }
    String s = merchantId + "|" + amount.toPlainString() + "|" + currency + "|" + caseId;
    return s.getBytes(StandardCharsets.UTF_8);
  }
}

File: services/buyer-simulator-ui/src/lib/canonicalForm.ts
/**
 * Browser mirror of {@code com.agentpay.buyer.keys.CanonicalForm}. Produces
 * the exact bytes the gateway verifier reconstructs from the JSON request
 * body — pipe-separated UTF-8, in the order
 * {@code merchant_id | amount | currency | case_id}.
 *
 * The {@code amount} field MUST match {@code BigDecimal.toPlainString()} on
 * the Java side: no scientific notation, trailing zeros preserved verbatim.
 * Callers therefore pass a pre-normalised string (e.g. "42.50") rather than
 * a JavaScript number — number-to-string conversion would drop trailing
 * zeros and break signature verification.
 */
export function canonicalForm(
  merchantId: string,
  amountPlain: string,
  currency: string,
  caseId: string,
): Uint8Array {
  const s = `${merchantId}|${amountPlain}|${currency}|${caseId}`
  return new TextEncoder().encode(s)
}
